libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -1,79 +1,141 @@
+import kotlinx.coroutines.withTimeoutOrNull
+import java.util.Date
+import kotlin.coroutines.resume
 
 /**
  * This class is responsible for refreshing the current direction route's traffic.
  * This does not support alternative routes.
- *
- * If the route is successfully refreshed, this class will update the [DirectionsSession].
  */
 internal class RouteRefreshController(
     private val routeRefreshOptions: RouteRefreshOptions,
     private val directionsSession: DirectionsSession,
-    private val tripSession: TripSession,
-    threadController: ThreadController,
+    private val currentLegIndexProvider: () -> Int,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
+    private val localDateProvider: () -> Date
 ) {
 
     internal companion object {
         internal const val LOG_CATEGORY = "RouteRefreshController"
     }
 
-    private val routerRefreshTimer = MapboxTimer(threadController).apply {
-        restartAfterMillis = routeRefreshOptions.intervalMillis
+    suspend fun refresh(routes: List<NavigationRoute>): List<NavigationRoute> {
+        return if (routes.firstOrNull()?.routeOptions?.enableRefresh() == true) {
+            val result = routes.toMutableList()
+            val routeToRefresh = result.first()
+            result[0] = tryRefreshingUntilRouteChanges(routeToRefresh)
+            result
+        } else waitForever("refresh isn't enabled")
     }
 
-    private var currentRequestId: Long? = null
-
-    /**
-     * If route refresh is enabled, attach a refresh poller.
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun restart(route: NavigationRoute, onRoutesRefreshed: (List<NavigationRoute>) -> Unit) {
-        stop()
-        if (route.routeOptions.enableRefresh() == true) {
-            routerRefreshTimer.startTimer {
-                refreshRoute(route, onRoutesRefreshed)
+    private suspend fun tryRefreshingUntilRouteChanges(
+        initialRoute: NavigationRoute
+    ): NavigationRoute {
+        while (true) {
+            val refreshed = refreshRoute(initialRoute)
+            if (refreshed != initialRoute) {
+                return refreshed
             }
         }
     }
 
-    /**
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun stop() {
-        routerRefreshTimer.stopJobs()
-        currentRequestId?.let {
-            directionsSession.cancelRouteRefreshRequest(it)
-            currentRequestId = null
+    private suspend fun refreshRoute(
+        route: NavigationRoute
+    ): NavigationRoute = coroutineScope {
+        val routeLegs = route.directionsRoute.legs()
+        require(routeLegs != null) { "Can't refresh route without legs" }
+
+        var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+        for (attempts in 0..2) {
+            timeUntilNextAttempt.await()
+            timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
```

Comment:
Why wrapping `delay` into `async`?

Reply:
Related to https://github.com/mapbox/mapbox-navigation-android/pull/5767#discussion_r883648854

Reply:
It helps me make stable interval of route refreshes. After I receive a result or an error or timeout from router I do `timeUntilNextAttempt.awat()`, so that interval doesn't depend on the time a router returns result. It may seem as too complex, but it's the best solution for our case where the router returns quick result doesn't call callback at all.

Reply:
I see, thanks :+1: 

What if we try to simplify the setup with something like this (or similar):
```
var elapsedAttemptTime = 0L
repeat(3) {
    delay(routeRefreshOptions.intervalMillis - elapsedAttemptTime)
    val time = System.nanoTime()
    val refreshedRoute = withTimeoutOrNull(routeRefreshOptions.intervalMillis) {
        refreshRouteOrNull(route)
    }
    elapsedAttemptTime = ((time - System.nanoTime()) / 1E6).toLong()
    if (refreshedRoute != null) {
        return@coroutineScope refreshedRoute
    }
}
```
Might be good to drop a comment as well.

Reply:
TBH, current implementation seems simpler for me. I also not sure if provided code is easy to test, it seems that I will need to mock `System.nanoTime()`, so time adjustment won't be as simple as `advanceTimeBy(30_000)`.

Is there any other option to make current approach more understandable? Maybe you can help me with naming? 🙂

Reply:
When do you think 
```
finally {
    timeUntilNextAttempt.cancel()
}
```
may be necessary?

Reply:
Cancellation is always needed here, so a `finally` clause simplifies the setup instead of having to define cancellation on each return statement.

Reply:
@LukasPaczos ,
You're may be considering cancellation(https://pl.kotl.in/XhJKKYqrt) or unexpected failure(https://pl.kotl.in/HwQkpcsjD). As you can see from examples those cases are already handled by structured concurrency. 
Do you suggest cancel explicitly because of some other case?

Reply:
No. I'm just trying to resolve the fact that you currently call
```
timeUntilNextAttempt.cancel()
```
twice, once for each return statement. If there was another espace point from the function, we'd have to manually cancel the coroutine again. This can be simplified with a `finally` statement instead.

Reply:
Thanks, now I see the point. Done.